### PR TITLE
ARROW-10119: [C++] Fix Parquet crashes on invalid input

### PR DESCRIPTION
**Pull request by pitrou**
Should fix the following issues (found by OSS-Fuzz):
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=25901
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=25933
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=25955
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=25957
- https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=25959

### CODE REVIEW

File: cpp/src/parquet/arrow/schema.cc [hunk 1]
```diff
@@ -834,16 +834,23 @@ Status SchemaManifest::Make(const SchemaDescriptor* schema,
...
   manifest->descr = schema;
   manifest->schema_fields.resize(schema_node.field_count());
+
+  // Try to deserialize original Arrow schema
+  RETURN_NOT_OK(
+      GetOriginSchema(metadata, &manifest->schema_metadata, &manifest->origin_schema));
+  // Ignore original schema if it's not compatible with the Parquet schema
+  if (manifest->origin_schema != nullptr &&
+      manifest->origin_schema->num_fields() != schema_node.field_count()) {
+    manifest->origin_schema = nullptr;
```
Comment (emkornfield) [inline cpp/src/parquet/arrow/schema.cc:851]: might be worth a log statement here?

File: cpp/src/parquet/column_reader.cc [hunk 2]
```diff
@@ -374,14 +381,10 @@ std::shared_ptr<Page> SerializedPageReader::NextPage() {
 
       page_buffer = decryption_buffer_;
     }
-    // Uncompress it if we need to
-    if (decompressor_ != nullptr) {
-      page_buffer = DecompressPage(compressed_len, uncompressed_len, page_buffer->data());
-    }
 
-    const PageType::type page_type = LoadEnumSafe(&current_page_header_.type);
+    current_page_type_ = LoadEnumSafe(&current_page_header_.type);
```
Comment (emkornfield) [inline cpp/src/parquet/column_reader.cc:385]: sorry, its not immediately obvious why thi sis becoming a member?

File: cpp/src/parquet/column_reader.cc [hunk 3]
```diff
@@ -391,10 +394,14 @@ std::shared_ptr<Page> SerializedPageReader::NextPage() {
         throw ParquetException("Invalid page header (negative number of values)");
       }
 
+      // Uncompress if needed
```
Comment (emkornfield) [inline cpp/src/parquet/column_reader.cc:397]: nit: might be better to name the method something like the comment.

**Comment by emkornfield** [comment]
A few small comments otherwise LGTM.

File: cpp/src/parquet/arrow/schema.cc (hunk 1 above)
Comment (pitrou) [reply cpp/src/parquet/arrow/schema.cc:851]: Or perhaps it should error out? I don't know.

File: cpp/src/parquet/column_reader.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/parquet/column_reader.cc:385]: Hmm, this was needed in a previous version of this patch, but I'll revert.

File: cpp/src/parquet/column_reader.cc (hunk 3 above)
Comment (pitrou) [reply cpp/src/parquet/column_reader.cc:397]: Will do :-)

**Comment by pitrou** [comment]
CI failure is just a docker cache issue.

**Pull request closed without merging**